src/main/java/seedu/address/model/tag/TagManagerImpl.java

```diff
@@ -0,0 +1,102 @@
+package seedu.address.model.tag;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import seedu.address.model.person.Person;
+
+/**
+ * Concrete implementation class of {@code TagManager} that uses a {@code HashMap}.
+ */
+public class TagManagerImpl implements TagManager {
+
+    private static final String MESSAGE_ERROR_PERSON_NOT_FOUND = "Person instance not found in tag manager!";
+
+    private final Map<Tag, Set<Person>> tagPersonSetMap;
+
+    public TagManagerImpl() {
+        tagPersonSetMap = new HashMap<>();
+    }
+
+    @Override
+    public Set<Person> getPersonsUnderTag(Tag tag) {
+        return tagPersonSetMap.get(tag) == null ? Set.of() : tagPersonSetMap.get(tag);
+    }
+
+    @Override
+    public Set<Tag> getTags() {
+        return tagPersonSetMap.keySet();
+    }
+
+    /**
+     * Removes all references in {@code tagPersonSetMap} to this {@code person} based on the {@code tag}s it has.
+     */
+    private void removeAllTagReferencesOfPerson(Person person) {
```

Comment:
consider adding the `NoSuchElementException` thrown to the method signature

Reply:
it is a runtime exception. I do not believe it is necessary. Especially since there should be no reason why this exception occurs at all if `Person` changes are done through AddressBook.

Reply:
noted. i mentioned this because `Parser.parse` (and all its implementations) contain an unchecked exception in it's method signature.